### ISSUE TITLE
Errors when upgrading to ruby 2.4 due to ruby bug

### ISSUE THREAD

**Issue by sgringwe** [librato/librato-rack#70]
It looks like this gem is impacted by a ruby bug, https://bugs.ruby-lang.org/issues/13107, from its use of `def_delegators`.

Would it be plausible to work around that ruby bug in this project?

**Comment by chancefeick** [comment]
Hi @sgringwe, thanks for the information. A couple of questions:

Which version of librato-rack do you have installed?

Do you have a custom instrumentation snippet to reproduce the issue, or is it sporadic like the bug report suggests? If so, can you provide an exception backtrace?

I'm unfortunately unable to reproduce on librato-rack 2.0.1, Ruby 2.4 and Rails 5.0.1.

**Comment by sgringwe** [comment]
We are using `librato-rack (1.1.0)`, `librato-rails (1.4.2)`, ruby 2.4, rails 5.0.1.

We also did not receive this error until we deployed into high throughput production environment which gave us ~1 error per 25,000 requests.

Example backtrace:

```
vendor/ruby-2.4.0/lib/ruby/2.4.0/forwardable.rb:228:in `timing': wrong argument type Integer (expected Proc) (ActionView::Template::Error)
    from vendor/bundle/ruby/2.4.0/bundler/gems/handshake_service-67ed91f455d4/lib/handshake_service/librato_helper.rb:21:in `timing'
    from app/services/toggles.rb:86:in `enabled?'
    from app/helpers/search_helper.rb:48:in `student_job_search_intro'
    from app/views/postings/dashboards/_student.html.haml:11:in `block in _app_views_postings_dashboards__student_html_haml__407538604791608508_117011240'
    from vendor/bundle/ruby/2.4.0/gems/haml-4.0.7/lib/haml/helpers.rb:368:in `block in capture_haml'
    from vendor/bundle/ruby/2.4.0/gems/haml-4.0.7/lib/haml/helpers.rb:608:in `with_haml_buffer'
    from vendor/bundle/ruby/2.4.0/gems/haml-4.0.7/lib/haml/helpers.rb:364:in `capture_haml'
    from vendor/bundle/ruby/2.4.0/gems/haml-4.0.7/lib/haml/helpers/xss_mods.rb:61:in `capture_haml_with_haml_xss'
    from vendor/bundle/ruby/2.4.0/gems/haml-4.0.7/lib/haml/helpers/action_view_mods.rb:45:in `capture_with_haml'
    from vendor/bundle/ruby/2.4.0/gems/actionview-5.0.1/lib/action_view/helpers/capture_helper.rb:153:in `content_for'
    from app/views/postings/dashboards/_student.html.haml:10:in `_app_views_postings_dashboards__student_html_haml__407538604791608508_117011240'
```

The line causing the error (https://github.com/strydercorp/handshake_service/blob/master/lib/handshake_service/librato_helper.rb#L21):

```
Librato.timing sanitized_name, time, options, &block
```

**Comment by chancefeick** [comment]
Wanted to follow up on this issue since there was some backchannel with support.

Potential workaround is to use the latest Ruby trunk, which includes patch that will be released in Ruby 2.4.1.

**Comment by abepetrillo** [comment]
Getting a similar error here: https://github.com/librato/librato-rack/blob/ebfb0ccd08e9ffb893974ae186dd104a7ec3ebcc/lib/librato/collector/aggregator.rb#L77

```
NoMethodError: undefined method `[]' for false:FalseClass
75       def measure(*args, &block)
76         options = {}
77         event = args[0].to_s
78         returned = nil
[GEM_ROOT]/gems/librato-rack-2.0.1/lib/librato/collector/aggregator.rb:77 :in `measure`
[GEM_ROOT]/gems/librato-rack-2.0.1/lib/librato/collector/group.rb:23 :in `measure`
```

**Comment by nextmat** [comment]
@abepetrillo - thanks for the report

The fundamental issue is that delegation in the standard library is broken in high-volume use in 2.4.0

We could stop using stdlib delegation (roll our own or drop use of delegation), but this would mean losing the performance benefits of its use in every version of ruby except 2.4.0.

For now our recommendation is either:
1) hold off on going up to 2.4.x until 2.4.1 is released
2) we may be able to provide a temporary branch to workaround this for specific cases until 2.4.1 is released

Feel free to reach out to customer success team at support@librato.com if we can help you further with this.

**Comment by sgringwe** [comment]
I can confirm that this is so far fixed when upgrading to ruby 2.4.1.

**Comment by chancefeick** [comment]
Thanks for the update @sgringwe 👍
